File: src/pratice/binarysearch/예산2512.java
package pratice.binarysearch;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 예산2512 {
    public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    public static int N, M;
    public static int[] A;

    public static void main(String[] args) throws IOException {
        input();
        Arrays.sort(A);
        pro();

    }

    private static void pro() {
        int L = 0;
        int R = A[N - 1];
        int res = L;

        while(L <= R){
            int mid = (L + R) / 2;
            if(isAble(mid)){
                res = mid;
                L = mid + 1;
            }else{
                R = mid  - 1;
            }
        }
        System.out.println(res);
    }

    private static boolean isAble(int mid) {
        int sum = 0;
        for (int i = 0; i < N; i++) {
            if(A[i] <= mid){
                sum += A[i];
            }else{
                sum += mid;
            }
        }
        return sum <= M;
    }

    private static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        A = new int[N];
        String[] split = br.readLine().split(" ");
        for (int i = 0; i < N; i++) {
            A[i] = Integer.parseInt(split[i]);
        }
        M = Integer.parseInt(br.readLine());
    }
}

File: src/pratice/binarysearch/예산2512.kt
package pratice.binarysearch

fun main() {
    val N = readln().toInt()
    val arr = readln().split(" ").map { it.toInt() }.toMutableList()
    val money = readln().toInt()

    arr.sort()

    lowerBound(arr, money)
}

private fun lowerBound(arr: MutableList<Int>, money: Int) {
    var L = 0
    var R = arr[arr.size - 1]
    var res = 0
    while (L <= R) {
        val mid = (L + R) / 2
        if (divideAble(arr, money, mid)) {
            res = mid
            L = mid + 1
        } else {
            R = mid - 1
        }
    }
    println(res)
}
private fun divideAble(arr: MutableList<Int>, money: Int, mid: Int): Boolean {
    var sum = 0
    for (num in arr) {
        if (num <= mid) {
            sum += num
        } else {
            sum += mid
        }
    }
    return sum <= money
}
